server/app/mutations/grid_certificates/authorize_domain.rb

```diff
@@ -9,32 +9,89 @@ class AuthorizeDomain < Mutations::Command
-      }
-      authz = get_authz_for_domain(self.grid, self.domain)
-      if authz
+      challenge = nil
+      case self.authorization_type
+      when 'dns-01'
+        debug "creating dns-01 challenge"
+        challenge = authorization.dns01
+        challenge_opts = {
+          'record_name' => challenge.record_name,
+          'record_type' => challenge.record_type,
+          'record_content' => challenge.record_content
+        }
+      when 'tls-sni-01'
+        debug "creating tls-sni-01 challenge"
+        challenge = authorize_tls_sni(authorization)
+        return unless challenge
+      end
+
+      if authz = get_authz_for_domain(self.grid, self.domain)
         authz.state = :created
-        authz.update_attributes(grid: self.grid, domain: self.domain, challenge: challenge.to_h, challenge_opts: challenge_opts)
+        authz.update_attributes(grid: self.grid, domain: self.domain, authorization_type: self.authorization_type, challenge: challenge.to_h, challenge_opts: challenge_opts)
       else
-        authz = GridDomainAuthorization.new(grid: self.grid, domain: self.domain, challenge: challenge.to_h, challenge_opts: challenge_opts)
+        authz = GridDomainAuthorization.new(grid: self.grid, domain: self.domain, authorization_type: self.authorization_type, challenge: challenge.to_h, challenge_opts: challenge_opts)
       end
 
       authz.save
       authz
     rescue Acme::Client::Error::Unauthorized
       add_error(:acme_client, :unauthorized, "Registration probably missing for LE")
     end
+
+    # Creates the self-signed cert from authorization object, puts that in vault and attaches it to the given service
+    # @return [Acme::Client::Resources::Challenges::DNS01, Acme::Client::Resources::Challenges::TLSSNI01] the actual challenge
+    def authorize_tls_sni(authorization)
+      challenge = authorization.tls_sni01
+      verification_cert = [challenge.certificate.to_pem, challenge.private_key.to_pem].join
+      secret_name = [LE_TLS_SNI_PREFIX, domain_to_vault_key(self.domain)].join('_')
+      tls_sni_secret = upsert_secret(secret_name, verification_cert)
+      if tls_sni_secret.nil?
+        add_error(:tls_sni_secret, :error, "Failed to store the needed tls-sni-01 secret")
+        return
+      end
+      debug "upserted tls-sni secret: #{tls_sni_secret.name}"
+      add_secret_to_service(tls_sni_secret, @lb_service) if @lb_service
+      challenge
+    end
+
+    # Adds the tls-sni secret to the service
+    def add_secret_to_service(secret, grid_service)
+      unless grid_service.secrets.index {|s| s['secret'] == secret.name}
+        info "adding tls-sni secret to service: #{grid_service.qualified_name}"
+        secrets = grid_service.secrets.map{ |s| {'secret' => s['secret'], 'name' => s['name'], 'type' => s['type']}}
+        secrets << {secret: secret.name, name: 'SSL_CERTS', type: 'env'}
+        outcome = GridServices::Update.run(grid_service: grid_service, secrets: secrets)
```

Comment:
My suggestion: don't modify the lb service secrets like this, instead, just `GridServiceAuthorization.create(linked_service: @lb_service)` and bump the `@lb_service` `revision`, `updated_at` fields. The `Rpc::ServicePodSerializer` can then pick up any `grid_service_authorizations` and serialize them as `SSL_CERT` secrets.

For bonus points, create a new `GridServiceDeploy` and return the deploy ID, so that the CLI can spin waiting for the challenge cert to get deployed.